src/main/java/org/folio/service/transactions/batch/BatchEncumbranceService.java

```diff
@@ -0,0 +1,113 @@
+package org.folio.service.transactions.batch;
+
+import org.folio.rest.jaxrs.model.Budget;
+import org.folio.rest.jaxrs.model.Encumbrance;
+import org.folio.rest.jaxrs.model.Transaction;
+import org.folio.rest.jaxrs.model.Transaction.TransactionType;
+
+import javax.money.CurrencyUnit;
+import javax.money.Monetary;
+import java.util.List;
+import java.util.Map;
+
+import static org.folio.rest.jaxrs.model.Transaction.TransactionType.ENCUMBRANCE;
+import static org.folio.utils.CalculationUtils.calculateBudgetSummaryFields;
+import static org.folio.utils.MoneyUtils.subtractMoney;
+import static org.folio.utils.MoneyUtils.sumMoney;
+
+public class BatchEncumbranceService extends AbstractBatchTransactionService {
+
+  @Override
+  public TransactionType getTransactionType() {
+    return ENCUMBRANCE;
+  }
+
+  @Override
+  public void updatesForCreatingTransactions(List<Transaction> transactionsToCreate, BatchTransactionHolder holder) {
+    createBudgetMapForTransactions(transactionsToCreate, holder.getBudgets())
+      .forEach(this::updateBudgetForEncumbranceCreation);
+  }
+
+  @Override
+  public void updatesForUpdatingTransactions(List<Transaction> transactionsToUpdate, BatchTransactionHolder holder) {
+    createBudgetMapForTransactions(transactionsToUpdate, holder.getBudgets())
+      .forEach((budget, encumbrances) -> updateBudgetForEncumbranceUpdate(budget, encumbrances,
+        holder.getExistingTransactionMap()));
+  }
+
+  private void updateBudgetForEncumbranceCreation(Budget budget, List<Transaction> encumbrances) {
+    CurrencyUnit currency = Monetary.getCurrency(encumbrances.get(0).getCurrency());
+    encumbrances.forEach(encumbrance -> {
+      double newEncumbered = sumMoney(budget.getEncumbered(), encumbrance.getAmount(), currency);
+      budget.setEncumbered(newEncumbered);
+    });
+    calculateBudgetSummaryFields(budget);
+    updateBudgetMetadata(budget, encumbrances.get(0));
+  }
+
+  private void updateBudgetForEncumbranceUpdate(Budget budget, List<Transaction> encumbrances,
+      Map<String, Transaction> existingTransactions) {
+    CurrencyUnit currency = Monetary.getCurrency(encumbrances.get(0).getCurrency());
+    encumbrances.forEach(encumbrance -> {
+      Transaction existingEncumbrance = existingTransactions.get(encumbrance.getId());
+      if (existingEncumbrance != null && isNotFromReleasedExceptToUnreleased(encumbrance, existingEncumbrance)) {
```

Comment:
if existing transaction is not exists when we are going to update budget based on it - it sounds line something happens, maybe lets write a warn log for this?

Reply:
Actually this can never happen (see `checkExistingTransactionsConsistency` and usage of `addTransactionToUpdate`). I am going to remove that test.